src/tree_walker.cpp

```diff
@@ -0,0 +1,179 @@
+#include "tree_walker.h"
+#include <iostream>
+
+#include <string> //feel free to delete later
+
+using namespace smt;
+using namespace std;
+
+namespace smt
+{
+
+//DELETE
+std::string vectorToString(vector<int> v){
+  std::string s;
+  for (int i:v){
+    s += std::to_string(i);
+    s += ", ";
+  }
+  return s;
+}
+
+pair<Term, vector<int>> TreeWalker::visit(Term & node)
+{
+  //iterates over children, for tracking which child on & path
+  int child_no;
+  //path for parts of the formula
+  vector<int> tree_path;
+
+  if (clear_cache_)
+  {
+    cache_.clear();
+
+    if (ext_cache_)
+    {
+      ext_cache_->clear();
+    }
+  }
+
+  pair <Term, vector<int>> out;
+  out.first = node;
+  out.second = tree_path;
+  if (query_cache(node, out))
+  {
+    // cache hit
+    return out;
+  }
+
+  //tree_path currently empty, process top node
+  visit_term(node, node, tree_path);
+  TermPairVec to_visit;
+  pair<Term, int> p1 (node, -1);
+  to_visit.push_back(p1);
+  child_no = 0;
+  for (auto ttt : node){
+    p1.first = ttt;
+    p1.second = child_no;
+    to_visit.push_back(p1);
+    child_no++;
+  }
+
+  Term t;
+  TreeWalkerStepResult res;
+  pair<Term, int> current_pair;
+  Term current_term;
+  pair<Term, int> pn;
+  while(to_visit.size())
+  {
+    current_pair = to_visit.back();
+    current_term = current_pair.first;
+    child_no = current_pair.second;
+    to_visit.pop_back();
+
+    if (child_no != -1){
+      tree_path.push_back(child_no);
+      visit_term(node, current_term, tree_path);
+      pn.first = current_term;
+      pn.second = -1;
+      to_visit.push_back(pn);
+      child_no = 0;
+      for (auto tt : current_term){
```

Comment:
```suggestion
      // adding children to `to_visit`
      for (auto tt : current_term){
```

Reply:
Is this necessary or is the line above the start of the for loop sufficient? Should that line be moved to within the for loop?

Reply:
I don't see any line above the loop currently, but anyway it is enough to have it a single time above the loop.